### PR TITLE
document and refactor the gc-stats code

### PR DESCRIPTION
**Pull request by gasche**
This is a series of refactoring commits for the "gc stats" part of the runtime code, starting with a documentation commit. It is meant to be reviewed commit-by-commit.

My goal is to be able to discuss and understand well what is going on with a refactoring-only PR, and then solve issues in followup PRs:
- We want "orphaned allocation statistics" to remove the current constraint that we have to reuse domain states (see the introduction of #10950); in fact, currently, unless I'm missing something, the orphaned shared-heap stats (free_pools.stats) are also not counted in the printed statistics. I have a patchset fixing both issues on the way, on to of the present PR.
- The computation of maximum stats values is broken right now (it will sum the maximums on all domains, which does not make sense), and we should also fix this. This would be a follow-up PR with some discussion, because we have to decide which behavior we want. (My proposal is to compute "total" values at sampling time during the stop-the-world, and have global state to keep track of the maximum of these total values.)
 
(Note: I believe that the computation of `top_heap_words` is also incorrect.)

Github suggests as potential reviewers @sadiqj, @Engil and @kayceesrk, and that sounds about right.

P.S.: I believe that there are not much tests for GC statistics in the current testsuite, this PR probably needs some.

### CODE REVIEW

**Comment by gasche** [comment]
Here is a prototype of allocation stats orphaning on top of the current PR: https://github.com/gasche/ocaml/compare/runtime-gc-stats...gasche:orphan-gc-stats . (I need to write tests for this before submission.)

File: runtime/caml/gc_stats.h
Comment (abbysmal) [inline runtime/caml/gc_stats.h]: We also collect statistics at domain termination IIRC.
Though domain termination induce a minor collection so it may be redundant?

File: runtime/gc_ctrl.c
Comment (abbysmal) [inline runtime/gc_ctrl.c:61]: I was thinking about moving the `stat_major_collections` stat at the same level as `minor_collections`, it feels odd that we have this one retrieved directly from `Caml_state` when we do not do the same for `minor_collections`.
Maybe this could be part of the refactor as well?

File: runtime/gc_stats.c
Comment (abbysmal) [inline runtime/gc_stats.c]: Ditto as for the header file: we also collect the samples at domain termination.

File: runtime/shared_heap.c
Comment (abbysmal) [inline runtime/shared_heap.c:293]: This bit seems to have been left out of the refactor in `pool_global_adopt_stats` ?

Comment (abbysmal) [reply runtime/shared_heap.c:293]: Actually it is now done in `caml_accum_heap_stats`, line 28, my bad.

File: runtime/gc_stats.c
Comment (abbysmal) [inline runtime/gc_stats.c:113]: Nice update to the comment. :-)

**Review by abbysmal — COMMENTED** [review]
This PR looks good to me and is a much welcome refactor of the gc stats code in the runtime.
One thing this highlights is that the top heap words computations is still pretty wrong at the moment.
I wonder if there's anything we can do to make it right? I will try to think about this.
For the orphaning of gc stats, feel free to get in touch with me if you want to discuss some details, or ping me when you think the change is ready for review (likely should be its own PR?).

**Comment by abbysmal** [comment]
> P.S.: I believe that there are not much tests for GC statistics in the current testsuite, this PR probably needs some.

Slightly related, we asked ourselves the same questions with @sadiqj about how to test for statistics altered by the GC machinery (specfically for the eventring PR #10964, not the GC stats themselves.).

The tricky part is how can we test for such things without creating a new class of flaky tests that we have to promote everytime adjustments are made to the GC.

Maybe it is somewhat fine? Although I think the testsuite already has some amount of such things (some testcases with location information we need to promote every time we shuffle some code around.).
I think the more recurrent promotes are, the more the human mind will forget to read thoroughly through these. :-)

Would be happy to discuss any idea about this specific topic.

**Comment by gasche** [comment]
Thanks for the excellent comments! All good suggestions, I'll work through them individually.

> One thing this highlights is that the top heap words computations is still pretty wrong at the moment.
> I wonder if there's anything we can do to make it right? I will try to think about this.

I haven't written any code, but my plan is as follows:
- define a category of "global stats" that make sense for the whole runtime, not individual domains
- move the "maximum" stats to these "global" stats
- update the global stats during STW sections

Currently the maximum statistics are computed by summing the maximum of all domains. This is completely wrong, because the domains may have reached their maximum at different times. Instead we should sum the *current value* of all domains (plus the orphaned stats, which are currently forgotten), and compare this to the historic (global) maximum value. Because this involves all domains, we do it during stop-the-world events, namely minor collections.

> For the orphaning of gc stats, feel free to get in touch with me if you want to discuss some details, or ping me when you think the change is ready for review (likely should be its own PR?).

Yes, my plan is to submit this as a separate PR. I worked on this already to convince myself that the refactoring was sensible / actually helpful, but I would like to work on testing the statistics as a priority.

> The tricky part is how can we test for such things without creating a new class of flaky tests that we have to promote everytime adjustments are made to the GC.

For the GC stats, my idea is to test not specific values, but weaker assertions. For example I would like to have a test that fires M domain which allocate N words each, and check that the total reported allocation is higher than M*N and lower than 2*M*N (a very tolerant window for extra allocations coming from the stdlib, etc.). I think that these would not be flaky -- if we design them right, they must always pass even after GC/runtime changes.

**Comment by gasche** [comment]
Re. summing "current values" but only at STW events:

- In theory this is an approximation, any statistics could vary between two STW events and sampling it irregularly may under-approximate the maximum value. This is a reasonable approximation if we assume that STW events occur often enough, and indeed this is the intended behavior of minor collections, although some workflows deviate from that -- then the stochastic approach of statmemprof is more appropriate, in the vein of [memprof-limits](https://gitlab.com/gadmm/memprof-limits).

- In practice, the GC stats we are talking about only concern the major heap, and in this case sampling at least at minor collections is a really good approximation, because those values rarely change outside minor collections. (It is still possible for the maximum at STW to be an under-approximation in the case where, between two minor collections, a major allocation finds a full pool and performs some sweeping work.)

Another approach would be to keep per-domain max values, but only during each window between two STW events (instead of max values over all history at today), by resetting the "observed maximums" to their current value at each STW event. Then it does make sense to sum those max values; they are a reasonable over-approximation (instead of under-approximation). This may be what was originally intended with the current code (but, as far as I can tell, it's not what it is doing). In fact we could even consider tracking both the under-approximation and the over-approximation if we wanted to be fancy, but I think that in practice the under-approximation is enough.

File: runtime/gc_ctrl.c
Comment (gasche) [reply runtime/gc_ctrl.c:61]: Done in a new commit, thanks for the suggestion!

File: runtime/caml/gc_stats.h
Comment (gasche) [reply runtime/caml/gc_stats.h]: I added an explicit mention, thanks.

File: runtime/gc_stats.c
Comment (gasche) [reply runtime/gc_stats.c]: I added a mention as well.

**Comment by gasche** [comment]
I pushed a rebased version taking @Engil's review comments into account.

**Comment by abbysmal** [comment]
Thanks :)

It appears that `weaklifetime_par` was impacted by the change, I cannot reproduce it on `trunk`, but on your branch it appears to fail one run out of three, on bytecode only.
It seems to trigger this assertion: https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/weak-ephe-final/weaklifetime_par.ml#L60

Maybe it is related to the change I suggested earlier? I will take a look.

(I will come back as well to answer your previous messages)

Edit:
Reverting the latest change seems to appease `weaklifetime_par`, there may be a subtle interaction this change breaks, I'm investigating.

**Comment by abbysmal** [comment]
I am thinking that this change actually creates an off-by-one versus before:
Now, the `major_collection` number is only updated at the end of a minor cycle (as this is when we collect the samples).

The `major_collection` stat is only updated at the end of the major cycle STW segment here: https://github.com/ocaml/ocaml/blob/trunk/runtime/major_gc.c#L1108

The minor collection part of the major cycle happens way before this stat is updated, meaning there is a off-by-one until the next minor segment, and this offset is caught by the testcase since it check for this very tight timing.

A bad fix would be, I assume, to just move the `major_collections` bumping to before the minor induced by the major cycle STW.
A better fix would involve one of the approaches you described earlier and having a global stat structure. (assuming we find that `minor/major_collections` belong to this category because of the stop-the-world segments involved in both.)

Maybe some people would find a per-domain number of collections to be a useful stat?
But then, it would be useful to have it actually tied to a domain, not a domain slot, ie, slot reuse kind of ruins the usefulness of such data point I think.

**Comment by abbysmal** [comment]
Point I forgot to make: in light of this finding, maybe this change can happen in a subsequent PR, when we address proper orphaning and/or global vs domain local stat story.

We should probably revert this change since this PR is otherwise pretty nice in my opinion.

**Comment by gasche** [comment]
Sorry it took me a bit of time to get back to this. This is tricky stuff!

Your very useful comments focus on what's going on specifically with the `major_collections` statistics. But from a more high-level point of view, the issue we observe here is just an instance of a more global behavior of the stats: during two minor GCs, only the statistics contribution from the current domain is reported accurately, and the part coming from other domains is computed using "stale" values coming from the last minor GC. So in a sense the statistics is "wrong", let's say "inaccurate".

In this particular example, we ask for `Gc.quick_stat().major_collections` at two different moments of the program (let's call them `gc1` and `gc2` below), and due to specific reasoning on ephemerons we know that there are at most two major collections apart, so `assert (gc2 <= gc1 + 2)`. But this reasoning is only correct if the values of `gc1` and `gc2` are accurate, which they are not in general in multi-domain settings. As you explain, there is a time window between the end of a major collection on another domain and the next minor collection where the `major_collections` count is inaccurate, and if `gc1` is computed during this time window the assert can fail.

I see several ways to fix the bug.
1. Your suggestion, generalized, is to keep track globally of statistics that count stopping-the-world events anyway, such as the (global) number of minor collections or major collections. This would fix the issue with the test, but it's an invasive change that I would rather keep for a further PR. And it does not make the problem go away that GC stats are sometimes inaccurate, and that we should think about how to adapt user programs that rely on them.
2. Of course, if you add a call to `Gc.minor()` before calling `quick_stat()`, then the stats are never inaccurate and the test works again. The test also becomes much faster, because its termination criterion is `while gccount () <= 5 do ... done`. (I think that the reason is that provoking many minor collections also accelerates the number of major GC slices we are counting? And not something related to the fact that the test would previously read an inaccurate `gccount ()`.) Maybe we should offer a `Gc.quick_stat ()` that may be inaccurate, and a `Gc.quick_enough_stat ()` that starts by doing a STW with sampling to guarantee accuracy. (In this specific test one can add an explicit `Gc.minor ()` not on all `quick_stat ()` calls but only on those just before sampling the `gc1` involved in the assert, and this suffices to make the issue go away.)
3. Another option would be to ensure that we sample GC stats not only in minor collections, but in all GC-related STW calls, or even all STW calls at all. Then we get the semantics that the stats are always "as accurate as possible", and the present problem should go away but also other potential problems with inaccurate stats.

**Comment by gasche** [comment]
@Engil I took your hint and removed the `major_allocations` commit for now. It's fairly subtle, we can make progress without it and deal with that in a later PR. I also added a Changes entry (I'm hoping for more PRs later that can be aggregated into a single entry.)

**Review by abbysmal — APPROVED** [review]

**Comment by abbysmal** [comment]
@gasche thank you for your answers.
The changes looks to me and I went ahead and approved the PR, I will merge it if you feel fine with it.

>  Another option would be to ensure that we sample GC stats not only in minor collections, but in all GC-related STW calls, or even all STW calls at all. Then we get the semantics that the stats are always "as accurate as possible", and the present problem should go away but also other potential problems with inaccurate stats.

I think this sounds reasonable, and I agree that option 1 is a bit invasive and probably better left as a thinking exercise for now.

> Maybe we should offer a Gc.quick_stat () that may be inaccurate, and a Gc.quick_enough_stat () that starts by doing a STW with sampling to guarantee accuracy. (In this specific test one can add an explicit Gc.minor () not on all quick_stat () calls but only on those just before sampling the gc1 involved in the assert, and this suffices to make the issue go away.)

I think this gets into the weird API territory, but there is a balance to be found in what people expects how `quick` things can be and how accurate it can get. (or how `inaccurate`).
From my time investigating the `core` testsuite I do remember some custom bits to get faster stats not using the GC APIs (I think involving accessing directly caml state fields, which is a bit hairy because these are domain-local and does not reflect overall statistics, and may get worse depending on the orphaning strategy we adopt.)

I think `quick_stat` as it stands, supplemented with your suggested idea at option 3, would provide a decently quick stat (I think we could survive, performance wise, to introduce gc sampling in all GC related STW events).
But there could be a faster path that is domain local, and would allow for fast domain specific lookups. (which, IIRC, is the usecase for the testsuite code I've seen, checking the number of words a function allocates as an assertion.).

**Comment by gasche** [comment]
> I will merge it if you feel fine with it.

Yes, please go ahead :-)

> But there could be a faster path that is domain local, and would allow for fast domain specific lookups. (which, IIRC, is the usecase for the testsuite code I've seen, checking the number of words a function allocates as an assertion.).

I agree. I think we should work together (and anyone interested) on a different API for stats, where users can explicitly request domain-local stats or global stats. (But we probably want to preserve some compatibility with code written against the current API.) This is longer-term than most of the refactorings discussed so far.

**Comment by gasche** [comment]
To summarize, here are things I think are worth doing in the short term, after this PR is merged, in no particular order:

- [x] (1) Properly handle orphaning of allocation stats. (This was my initial motivation for doing this work, and I have an experimental patchset to turn into a PR.) (Edit: #11047, now merged.)
- [ ] (2) Fix the computation of maximum values, probably by introducing a proper notion of "global stats".
- [ ] (3) Experiment with sampling after more STW events, possibly all of them.

(2) would get us moving in the direction of a better API, but it's focused on fixing a bug.

**Pull request merged**
